### PR TITLE
Add relation option "createForeignKeyConstraints"

### LINKED ISSUE #3120: Option to disable foreign keys creation
**Issue type:**

[x] question
[ ] bug report
[x] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

Is there a way to disable foreign keys creation? I searched a bit and couldn't find it.

If not, is there a chance such option can be implemented? Maybe I can give it a try if you point me in the right direction. (:

### PR DESCRIPTION
**Pull request by alpharder**
### Description of change

Fixes #3120 by adding a new relation option "createForeignKeyConstraints"

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [ ] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Review by nebkat — CHANGES_REQUESTED** [review]
From what I can tell `ForeignKeyMetadata`, other than to create `TableForeignKeys`, is only used to pass the `joinColumns` and `inverseJoinColumns` to this function:

https://github.com/typeorm/typeorm/blob/c4a36da62593469436b074873eba186f0f8b990d/src/metadata/RelationMetadata.ts#L490-L500

This function is only called in two places:

https://github.com/typeorm/typeorm/blob/c4a36da62593469436b074873eba186f0f8b990d/src/metadata-builder/EntityMetadataBuilder.ts#L195
https://github.com/typeorm/typeorm/blob/c4a36da62593469436b074873eba186f0f8b990d/src/metadata-builder/EntityMetadataBuilder.ts#L130-L131

Instead of having `ForeignKeyMetadata`s that aren't created in DB, why don't we eliminate them at this stage and pass `joinColumns` and `inverseJoinColumns` into a separate function (perhaps rename it to `registerJoinColumns()`) and push foreign keys into `relation` the same way they are pushed into `entityMetadata` above? After all if foreign keys aren't in the database, should we really ever call them foreign keys at all?

**Comment by acepace** [comment]
@alpharder what help do you need to finish this PR? I'm also blocked on this

**Comment by nebkat** [comment]
@acepace I've described some changes I that I think need to happen, but it is still usable in its current form. If you urgently need this I suggest you fork and cherry-pick what is here. Otherwise if @alpharder has not begun you could try to make those changes yourself and submit a new PR, but it's best to wait a while first. You can :+1: the PR to give it more attention.

**Comment by alpharder** [comment]
@acepace Well, this is 100% usable as is, because the patched version of TypeORM is used at our production for like 6 months already.

However, I haven't dived too much into TypeORM's internals because we've in the middle of migrating to MikroORM and most probably @nebkat 's comments should be taken into account before merging this into the mainline. 

I think I will finish with this until the end of the week, but another question is – how time much will it take to merge it after changes I'll make.

If you need to use it right now I suggest you to:
1) Install the latest version of TypeORM (or update an existing one) and commit changes; 
2) Clone my branch containing the fix;
3) Build it manually;
4) Put the build result files into corresponding typeorm directory at your project's node_modules
5) Use patch-package to make a permanent patch

**Comment by acepace** [comment]
That's awesome.
I'll see about fixing @nebkat comments and if so will open a joint PR (like, joint commit history, not sure how to give us both permissions).

**Comment by pleerock** [comment]
This PR looks like a good start to fix #3120 . Does anyone want to finish it? If no, I'll have to close it for now.

**Comment by Saniol** [comment]
> This PR looks like a good start to fix #3120 . Does anyone want to finish it? If no, I'll have to close it for now.

I finished changes and (as I couldn't update this one) submitted them in new PR - #7277

**Comment by nebkat** [comment]
Fully completed in #7277, discussion can move there.

**Comment by pleerock** [comment]
Closing in favor of #7277

**Pull request closed without merging**
